From 0b79bda3d6c72a4260e55572e3cb3bfadf782f12 Mon Sep 17 00:00:00 2001
From: Billy Tsai <billy_tsai@aspeedtech.com>
Date: Wed, 20 Oct 2021 18:42:33 +0800
Subject: [PATCH 2/3] peci: aspeed: Add aw_fcs when executing peci write

The aspeed peci hardware can auto replace the last data with awfcs. Thus
the driver need to enable this feature when the command code is write.

Signed-off-by: Billy Tsai <billy_tsai@aspeedtech.com>
Change-Id: Ieda4bfaa7203d149fde9ac000a8aa260b562a88c
---
 drivers/peci/peci_aspeed.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/drivers/peci/peci_aspeed.c b/drivers/peci/peci_aspeed.c
index d43bf83a9e..f7e9e096b8 100644
--- a/drivers/peci/peci_aspeed.c
+++ b/drivers/peci/peci_aspeed.c
@@ -208,6 +208,23 @@ static int peci_aspeed_enable(const struct device *dev)
 	return 0;
 }
 
+static int peci_aspeed_need_aw_fcs(enum peci_command_code cmd_code)
+{
+	switch (cmd_code) {
+	case PECI_CMD_WR_PCI_CFG0:
+	case PECI_CMD_WR_PCI_CFG1:
+	case PECI_CMD_WR_PKG_CFG0:
+	case PECI_CMD_WR_PKG_CFG1:
+	case PECI_CMD_WR_IAMSR0:
+	case PECI_CMD_WR_IAMSR1:
+	case PECI_CMD_WR_PCI_CFG_LOCAL0:
+	case PECI_CMD_WR_PCI_CFG_LOCAL1:
+		return 1;
+	default:
+		return 0;
+	}
+}
+
 static int peci_aspeed_transfer(const struct device *dev, struct peci_msg *msg)
 {
 	struct peci_aspeed_data *peci_data = DEV_DATA(dev);
@@ -232,6 +249,10 @@ static int peci_aspeed_transfer(const struct device *dev, struct peci_msg *msg)
 	peci_header.fields.enable_aw_fcs_cycle = 0;
 	peci_header.fields.write_data_length = msg->tx_buffer.len;
 	peci_header.fields.read_data_length = msg->rx_buffer.len;
+	if (peci_aspeed_need_aw_fcs(msg->cmd_code)) {
+		/* hardware will auto replace the last data with awfcs */
+		peci_header.fields.enable_aw_fcs_cycle = 1;
+	}
 	peci_register->peci_header.value = peci_header.value;
 	if (msg->tx_buffer.len) {
 		/* set write data */
-- 
2.25.1

